### DISCUSSION TITLE
Benötige ein paar Infos zum Bau des Moduls

### DISCUSSION THREAD

**Discussion by paxrsi** [dewenni/ESP_Buderus_KM271#106] [Q&A] [answered]
Hallo alle,
Ich interessiere mich für die Option 2 mit Buderus KM271 Modul. Das Modul habe ich auch schon gekauft. Ich habe mich für diese Option entschieden, weil ich möglichst wenig Risiko eingehen will, dass die Heizung bei dem Gebastel von mir evtl. einen Defekt bekommt. Im Grunde wäre mir das egal, aber bei dem neuen Heizungsgesetz kann ich das nicht gebrauchen, dass ich zur Unzeit plötzlich das ganze Haus auf Wärmepumpe umstellen muss. Irgendwann und kontrolliert mit Vorbereitung gerne, aber so ist mir das halt zu heikel.

Ich traue mir das auch zu, das alles selbst zu bauen. Allerdings komme ich eher aus dem Softwareumfeld und habe quasi nichts mit Mikrokontrollern und Kabellöten zu tun. Grundsätzlich sehe ich da kein Problem, aber der Teufel steckt halt im Detail und da will ich lieber zuerst fragen, bevor ich was falsch mache.

Zunächst brauche ich etwas Unterstützung, was die Teileliste angeht:

1. Bezgl. des MAX3232 hatte ich an den hier gedacht. Ist der ok? https://www.amazon.de/ANGEEK-Serial-Converter-Connector-MAX3232/dp/B07ZDK4BLH
2. Das serielle Kabel ist ein gekreuztes Kabel, richtig? Also ein Nullmodemkabel? Die meisten Kabel, die ich so finden konnte, sind entweder viel zu lang oder 1:1 Kabel, die dann noch ein kleines Nullmodemmodul vorgeschaltet bräuchten. Da bin ich mir dann aber nicht mehr sicher, ob das wirklich alles noch so funktioniert. Ich überlege schon, mir ein einfaches Kabel selbst zu löten. Habt ihr Vorschläge für ein passendes (kurzes) Kabel?
3. Außerdem frage ich mich, ob nicht auch die 5V an den MAX3232 und wohl von dort weiter an den ESP durchgereicht werden müssen. In der "Example configuration" sind nur TXD, RXD und GND beschrieben. Aber die Module brauchen doch auch eine Betriebsspannung, oder? Die müsste dann über das serielle Kabel vom KM271 kommen. Richtig?
4. Zum ESP: Ich habe einen AZDelivery ESP32 Dev Kit C V4 hier und einen ESP8226-12F Mini Board Node MCU. Passen die? Ich würde aber auch https://www.amazon.de/AZDelivery-Bluetooth-Internet-Entwicklungsboard-kompatibel/dp/B08BTLYSTM kaufen. Das ist wohl der auf dem Bild. Auf ein paar Euro kommt es nicht an.
5. Habt ihr Empfehlungen für ein Gehäuse? Der Aufbau in dem Bild gefällt mir schon ganz gut. Leider kann ich die Abmessungen nicht abschätzen.
6. Wird die Lochrasterplatine zurechtgeschnitten oder ist die direkt in der richtigen Größe gekauft? Sind die Befestigungsschrauben dabei oder kommen die extra?
7. Wie wird die Verdrahtung auf der Rückseite der Lochrasterplatine ausgeführt?
8. Ist beim USB-C Port nur das Gehäuse aufgeschnitten und das schwarze hinter dem USB-C Anschluss auf dem Bild ist die Befestigungslippe des Gehäuses (so wie rechts vom Gehäuse)?

Dann zu dem Ölmeter:
Laut Beschreibung kann man auch einen Ölzähler direkt anbinden. Leider steht nicht viel dazu, wie das geschieht. Da würden mich die Details interessieren, damit ich das gleich mit realisieren kann. Ich habe vom Heizungsbauer ein Angebot für einen Braun HZ3 und einen HZ5 DR. Da ist preislich quasi kein Unterschied. Daher tendiere ich zu dem HZ5 DR. Nur ist mir nicht klar, wie ich von dem Reed-Kontakt des HZ5 DR den Impuls in den ESP bekomme und welche Teile ich dafür benötige.

Vielen Dank für eure Unterstützung!

**Comment by the78mole** [comment, marked as answer]
Hallo Pax,

gerade weil Du sagst, dass Du eher der Software-Typ bist, würde ich Dir diese Option eher nicht empfehlen. Viel günstiger wirst Du mit dem Eigenbau nicht kommen und Du kannst wahrscheinlich schneller was kaputt machen, als mit meiner [KM271-WiFi](https://github.com/the78mole/km271-wifi). Bisher sind davon 300 Stück im Umlauf und ich hab bisher nie von defekten Steuerungen erfahren. Das Einizge, was Dir passieren kann ist, dass die Buderus ein zu schwaches Netzteil hat und Du meine Erweiterung dann auf USB-Power-Supply umstellen musst (sind geschätzt 10% der Steuerungen von betroffen). Das merkt man, wenn der ESP öfter mal einen Reset macht. Kaputt geht dabei nichts. Aber auch die Umstellung ist ganz einfach: Jumper umstecken, USB-Netzteil anschließen, fertig.

Falls Du das tatsächlich selbst aufbauen möchtest, kannst Du ja auch mal im [Mikrocontroller-Thread zum Thema](https://www.mikrocontroller.net/topic/141831?page=single) graben und fragen. Da gibt es deutlich mehr Leute, die selbst schon gebastelt haben und Dir da schneller weiterhelfen können, als die relativ kleine Community, welche dieses GitHub-Projekt beobachtet. Ich sehe da sehr viel Support-Aufwand auf mich (oder dewenni) zukommen und den kann ich nicht leisten (Familie, Arbeit, Haus-Sanierung, Ethernet-Erweiterung für meine KM271-WiFi, Verein,...)

Aber ich möchte trotzdem ganz kurz auf Deine Fragen eingehen:

1. Sollte funktionieren

2. Das kann ich Dir nicht beantworten. Eine bekannte Schwäche von RS232 ist, dass das jeder macht, wie er möchte. Aber kaputt kann nichts gehen, wenn Du es falsch verbindest. Die RS232-Treiber sind kurzschluss- und verpolungsfest. Etwas schwieriger dürfe es werden, festzustellen, ob es funktioniert. Die Buderus ist nicht sehr geschwätzig, wenn man nicht die richtigen Bytes zu ihr schickt.

3. Die Spannungsversorgung willst Du nicht von der Buderus nehmen. Die ist schon bei meinem Modul am Limit und bei Dir kämen noch zusätzliche Komponenten (RS232-Treiber auf Buderusseite, RS232-Treiber auf ESP-Seite) dazu. Also nein, Du musst den Rest der Schaltung separat versorgen (siehe Antwort 10).

4. Nimm den ESP32, der 8266 ist der ältere, mit weniger Speicher, langsamen Prozessor,.. Ich bin mir ziemlich sicher, dass dewennis Firmware nicht auf den ESP8266 draufpasst. Ich bin mir aber aktuell leider auch nicht sicher, ob einer mit 4MB Flash ausreicht oder ob Du einen mit 8 MB brauchst. Im Zweifel einfach einen mit 8 oder 16 MB nehmen.

5. Da hab ich keine Empfehlung. Abmessungen hingegen kannst Du Dir ganz einfach abzählen. Das Raster einer Lochrasterplatine ist genau 2,54mm von Lochmitte zu Lochmitte. Ich drucke mir sowas mittlerweile immer selbst. Geht schneller und ist günstiger, als stundenlang irgendwelche Reichelt- oder Amazon-Seiten zu durchwühlen. Und meist mache ich auch gleich eine eigene Platine.

6. Lochraster gibt es in allen erdenklichen Größen. Standard ist 100x160 mm. Mittlerweile bekommt man aber bei den asiatischen Händlern meist irgendwelche absurde Größen. Üblicherweise kramt man die Restekiste nach einem Stück durch, das passen könnte bzw. etwas größer ist und längt es dann passend ab. Kommt auch drauf an, welches Gehäuse gerade in der Kruschkiste verfügbar ist ;-) Bzgl. Schrauben: Meist sind die nicht dabei (und die Löcher in der Lochrasterplatine musst Du ebenfalls selbst bohren). Für mich haben sich da die Schrauben bewährt, mit denen früher Festplatten und Mainboards in Desktop-PCs festgeschraubt wurden. Aber auch nur deshalb, weil ich davon noch immer gefühlt 10 kg in einer Box habe. Und viel wichiger: Sie passen meist ;-)

7. Und da kommen wir zu dem Punkt, der etwas schwieriger werden dürfte... 
 
Das "grundsätzliche Wie": Entweder man nimmt isolierte kleine Drähte (oder Kupferlackdraht) und verbindet direkt die passenden Anschlüsse oder man biegt nicht-isolierte Drähte säuberlich in 90° Winkeln und punktet sie dann fest (beides geht nur bei Punkrasterplatinen). Manchmal bietet es sich auch an, Steifenrasterplatinen zu verwenden, das ist hierfür aber schwierig. Grundsätzlich ist die erste Option (mit dünnen, kunststoff-isolierten Drähten) am einfachsten, weil man alleine mit dem Schaltplan eine einfache Verdrahtung schnell hinbekommt. Nimmt man dann noch ein wenig Farbe mit rein (z.B. rot für Versorgung, schwarz für Masse), dann wird es sogar halbwegs nachvollziehbar. 

"Das konkrete Wie": Ich denke, die Lochraster ist hier tatsächlich nur aus mechanischen Gründen, damit man das ESP-Board im Gehäuse festschrauben kann. Die Leitungen gehen höchstwahrscheinlich direkt zu den nächstegelegenen Pins des ESP-Boards.

Nochmal generell:
Grundsätzlich ist es aber einfacher, es erst mal fliegend mit Dupont-Wires aufzubauen. Da finden sich die Fehler schneller und sind einfacher zu beheben. Ein Beispiel für einen sogenannten Drahtigel siehts Du hier:

![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/7723396/57f9c546-62ae-4a72-99fb-ca7885b4039d)

10. Da ist einfach das Gehäuse aufgeschnitten, damit man den ESP und den RS232-Treiber über USB versorgen kann. Programmieren geht darüber bei den meisten Boards auch.

So, ich hoffe, das hat Dir weitergeholfen... Trotzdem meine Empfehlung: Nimm das KM271-WiFi von mir. Das ist in 2 Minuten installiert und funktioniert out-of-the-box. Wenn Du es bei mir direkt bestellst, dann ziehe ich gerne noch die Tindie-Provision (ca. 10%) ab ;-)

PS: Den Ölzähler kannst Du Dir eigentlich sparen. Wenn Du die empfohlene Düse verbaut hast, dann weißt Du genau, wie viel Öl pro Stunde da durchgepresst wird (bei 21 kW-Brenner mit Standard-Düse sind es z.B. 2,1 Liter). Das multipliziert mit dem Betriebsstundenzähler und Du hast die absolute Ölmenge Deines Brenners. Wenn Du unbedingt den Oilmeter anschließen möchtest, dann geht das auf meiner neuesten Karte auch sehr einfach. Dafür hat die mittlerweile ein Sensor-Wire-Terminal. Da kann man dann einfach die beiden Leitungsenden des Ölmessers abisolieren und einstecken. Anschließend die Pins in dewennis Weboberfläche konfigurieren, fertig.

Beste Grüße,
Daniel

**Comment by paxrsi** [comment]
Hallo Daniel,

vielen Dank für deine ausführliche Antwort! Wahrscheinlich hast du Recht. Ich sollte mir wohl besser ein Projekt mit weniger Schadenspotenzial für meine ersten ESP-Gehversuche aussuchen...

Wie bestelle ich denn "direkt bei dir"? Ich bräuchte auch noch etwas Unterstützung, welche Optionen ich bei der Bestellung sinnvollerweise wähle, bzw. ob es Sinn macht schon mal vorbeugend eine Option mitzubestellen und worin sich die beiden Lösungen in der Anwendung unterscheiden. Hauptsächlich will ich die Heizung in HomeAssistant integrieren und über das Jahr hinweg (insbesondere den Winter) ein Verbrauchsprofil der Heizung aufzeichnen:
- Die Debug-PINs brauche ich nicht. Wenn die nicht schaden (z.B. unnütz Strom ziehen und dann reicht die Spannung nicht), kann ich sie aber mitbestellen. Wer weiß, wozu es mal gut ist.
- Gleiches gilt für den Abgas Sensor Header (ich nehme an, auf deutsch ist das der Anschluss für einen Abgas-Sensor, der noch extra hinzugekauft werden muss. BTW: Wofür braucht man den?)
- Was ist mit Sensor Cable Header gemeint?
- Brauche ich eine spezielle Option, falls ich eines Tages einen externen Ölzähler anschließen möchte?
- Bei der Firmware weiß ich leider gar nicht, was besser wäre... Das Primärziel ist die Heizung in HomeAssistant einzubinden. Eine ordentliche Weboberfläche wäre schön, kann aber zurückstehen, wenn das für HomeAssitant irgendwie nachteilig wäre. ESPHome habe ich in mein HomeAssistant bisher noch nicht eingebunden. Aber das kann ich ja noch machen. Einige smarte Tasmota-Steckdosen habe ich mit MQTT angebunden und einige Zigbee-Geräte mit Zigbee2MQTT.

Zur Düse im Brenner kann ich nichts sagen. Ich vermute, das ist die Düse, die einfach mit dem Gerät mitkam. Im Zweifel müsste ich den Heizungsbauer fragen, ob der noch Unterlagen hat, was er damals verbaut hat. Der Brenner hat 17 kW. Wenn man das über die Brenndauer annähern kann, wird mir das denke ich reichen. Ich brauche das nicht auf den Liter genau.


vielen Dank und viele Grüße
Rainer

**Reply by the78mole** [reply]
Hallo Rainer,

einfach eine Email mit Adresse an me(at)the78mole.de. Die Optionen kannst Dir [bei Tindie](https://www.tindie.com/products/the78mole/buderus-km271-wifi-replacement/) raussuchen und dann einfach einen Screenshot der Drop-Down-Auswahl anheften. Eine genauere Beschreibung (z.B Schaltplan) findest Du [hier](https://the78mole.de/projects/km271-wifi-howto/), Details zur Entwicklung der KM271-WiFi [hier](https://the78mole.de/reverse-engineering-the-buderus-km217/). Die kompletten Design-Files sind in meinem [Github-Repo](https://github.com/the78mole/km271-wifi).

Ja, an den Abgassensor kann man einen 100K-NTC anschließen. Der muss natürlich die Abgastemperaturen aushalten (insbesondere die Anschlussleitung ist hier die schwache Stelle). Ein möglicher Sensor ist [dieser](https://www.sensorshop24.de/einschraubtemperaturfuehler-mit-g1-4-gewinde) mit folgender Auswahl:
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/7723396/4abad58f-039a-480c-86de-eebf0c205808)

Die Abgastemperatur ist natürlich einerseits ein Hinweis, ob der Brenner überhaupt funktioniert, andererseits kann man damit natürlich gut auf die Effizienz der Verbrennung zurückschließen. BTW: Der 100K-NTC hat nicht exakt die gleiche Kennlinie wie der Original-Fühler. Bei 25°C stimmen sie exakt überein (100K), bei ca. 150°C zeigt die Buderus mit dem obigen Sensor aber nur etwa 130°C an. Das liegt daran, dass ein NTC (speziell dessen Kennlinie) nicht normiert ist, wie es z.B. bei Platin-Sensoren der Fall ist. Aber ich denke, die Genauigkeit ist ausreichend und wichtig ist letztlich der relative Wert. Wenn der Schlotfeger mal die Temperatur mit seinem Gerät misst, kann man ja spicken und seinen Wert im Home Assistant kalibrieren ;-)

Über den Sensor Cable Header kannst Du nahezu beliebige Sensoren oder Aktoren recht einfach anschließen (da musst Du dann nicht löten). Durch kleine Änderungen z.B. des Widerstandsteilers oder der Kondensatoren kann man den Eingang auf diverseste Signale anpassen...
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/7723396/9c18a6d4-5245-45cf-a8aa-66aeaa34b902)

Also die ganzen Optionen schaden nicht und sind ja auch preislich nicht wirklich eine große Last.

Beispiel: Wenn Du einen Ölzähler hast, der Impulse über einen potenzialfreien Kontakt (Schalter) ausgibt, dann schließt Du den einfach an Pin1 (GND) und Pin2 (CH7) an und bekommst ein Active-Low-Signal (ADC1_CH7) an den IO35 des ESP.
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/7723396/d9f07b12-6e68-44f3-8d40-62ef414b6397)

Dann kannst Du im ESP diesen einfach als invertierten Eingang (da Active-Low) konfigurieren und damit einen Zähler beschicken, der die Ölmenge hochzählt. Anmerkung: Der Pin wird dabei als GPIO verwendet, nicht als ADC.

Zum Brenner und Ölverbrauch... In meinem Ladekreis des Ölkessels habe ich einen Wärmemengenzähler verbaut, der mit die tatsächliche Wärmeleistung des Brenners genau misst. Der zeigt in der Spitze ca. 16 kW an. Der Brenner hat eine Nennleistung von 21 kW.

So, nun noch zur Firmware. Sowohl ESPhome, als auch dewennis FW sind innerhalb weniger Minuten eingerichtet. Mit ESPhome hast Du das auch sofort im HomeAssistant, bei dewenni musst Du halt noch MQTT sauber konfigurieren. Allerdings musst Du bei ESPhome selbst noch entscheiden, welche Sensoren, Numbers,... Du im HA haben willst. Du solltest Dir für die initiale Konfiguration durchaus mal ein paar Stunden einplanen. Speziell, wenn Du das YAML-Format von ESPhome noch nicht kennst. Idealerweise schaust Du die die Konfigs in meinem [Firmware-Repository](https://github.com/the78mole/esphome_components) mal an.

Ich würde Dir empfehlen, einfach mal mit dewennis Firmware anzufangen. Seit der neuesten Release unterstützt dewenni sogar Home Assistant Discovery. Damit ist die Einrichtung dann sogar noch einfacher als mit meinem ESPhome.

Außerdem: Mit einem simplen USB-TTL-seriell-Kabel (3,3V!!!) kannst Du das in 5 Minuten umprogrammieren. Ein Video zur Programmierung findest Du [hier](https://www.youtube.com/watch?v=h_pQlpXaQ1I&t=810s).

Sorry, ist doch wieder etwas länglicher geworden 😂

Grüße,
Daniel

**Comment by joeflower** [comment]
Du hast das wundervolle Webinterface von dewenni vergessen. Da braucht es nicht einmal zwingend einen Broker dahinter. (obwohl schon sinnvoll). Einrichtung im IOBroker in 5 Min. MQTT konfigurieren und die Werte werden alle erkannt, so das man damit "spielen" kann. Ich habe zwar auch HA laufen, habe aber die Discover Funktion noch nicht getestet. Ich wäre definitiv für die Soft von dewenni.

**Comment by the78mole** [comment]
Ja, das sowieso. Aber das war nicht Teil der Fragen in diesem Thread, weil der explizite Wunsch war, die Buderus in den Homeassistant einzubinden. Aber in jedem Falls ist das WebUI krass gut. 😍

PS: Tatsächlich ist das Verhältnis der verschickten KM271-WiFi deutlich zugunsten von ESPhome... Warum, weiß ich nicht genau. Seit das Discovery-Feature da ist, überlege ich selbst, umzusteigen. Dann müsste ich aber den Support für ESPhome droppen und das ist leider keine Option (zumindest, bis ich meine Ölheizung rauswerfe, aber dann brauche ich auch nicht mehr wechseln 🤣)

**Comment by paxrsi** [comment]
Nochmal Danke an alle Antworten!
Für diejenigen, die es interessiert wie es ausgegangen ist: Ich habe seit einiger Zeit ein KM271-WiFi mit der dewennis FW installiert und an Home Assistant angebunden. Überwiegend funktioniert das zuverlässig. Für die offenen Fragen dazu würde ich dann lieber einen neuen Thread mit passenderem Titel eröffnen.
Da mein Anliegen damit gelöst ist, würde ich gerne diesen Thread hier schließen - mal schauen, ob mir das gelingt.
